Introduce XXH_SIZE_OPT and XXH_NO_STREAM

easyaspi314 opened the pull request:
- `XXH_SIZE_OPT` is a value from 0-2 which indicates how much xxHash should care about code size, default 1 for -Os/-Oz, default 0 otherwise. It chooses a set of defaults. 
   - `==1`: conservative size optimizations (basically the same as what we were already doing with -Oz)
   - `==2`: performance may cry
 - `XXH_NO_STREAM` disables the streaming API.
 - These two interact: if `XXH_SIZE_OPT == 2` and `XXH_NO_STREAM` is not defined, XXH32 and XXH64 use the streaming API for single shot
    - TODO: apply this to XXH3 as well

easyaspi314 committed:
Introduce XXH_SIZE_OPT and XXH_NO_STREAM

 - XXH_SIZE_OPT is a value from 0-2 which indicates how much xxHash
   should care about code size, default 1 for -Os/-Oz, default 0
   otherwise
 - XXH_NO_STREAM disables the streaming API.
 - These two interact: if XXH_SIZE_OPT == 2 and XXH_NO_STREAM is not
 defined, XXH32 and XXH64 use the streaming API for single shot
    - TODO: apply this to XXH3 as well

Cyan4973 commented:
> if XXH_SIZE_OPT == 2 and XXH_NO_STREAM is not defined, XXH32 and XXH64 use the streaming API for single shot

How much binary size is saved by this optimization ?

> TODO: apply this to XXH3 as well

Do you want to do it as part of this PR ?

Cyan4973 commented on xxhash.h:
```diff
@@ -1591,6 +1629,17 @@ XXH3_128bits_reset_withSecretandSeed(XXH3_state_t* statePtr,
+
+/*!
+ * @def XXH_NO_STREAM
+ * @brief Disablea the streaming API.
```
nit: Disable

easyaspi314 committed:
[sizeopt] Reroll midrange on size opt, disable prefetch

It seems that using the streaming API is not beneficial on GCC so it
isn't worth the complexity.

easyaspi314 committed:
Minor typo fix

easyaspi314 commented:
I checked it. While Clang seems to do very slightly better doing streamed single shot XXH3, GCC emits larger code. However that might be AArch64 Clang's identical code outliner copying the stack setup.

So, not worth the 40-50 lines to implement it.

XXH3 has a lot more code reuse than XXH32/64 so it isn't too surprising that this is the case. (Which is why I expect some more improvement once I clean up the modern XXH32/XXH64 refactor). 

I also rerolled the midrange loops on >= 1.

Cyan4973 commented:
While I believe that the objectives of the PR are fine,
I sometimes worry about the added maintenance complexity.

I see nothing "wrong" in this PR.
Just another set of build variables and `#ifdef` to maintain,
and while they only add a little bit of complexity,
there is a sense that little by little, as PR stack up, following the code is getting more difficult.

And unfortunately, I also have no simple counter proposal.
Sometimes, making the code more maintainable requires sweeping non-local changes for the sake of small complexity benefits.

So take it more as a "general feeling" when building future evolutions (not something specific to this PR).
I think code maintenance is a topic we must keep in mind and try to improve.
